### ISSUE TITLE
joins does not works with multiple levels

### ISSUE THREAD

**Issue by wbotelhos** [shioyama/mobility#266]
First of all, thank you for the great gem.

I found a case that joins does not works. When attributes is on the root of hash, it works, but with join we need a second level attributes, so the joins is not triggered.

## Context

I'm using backend key/value with `query_order` branch on MySQL.

## Behaviors

This code works with `slug` on root of the hash condition:

```
Category.i18n.joins(:category_profiles).where(slug: "the-title").to_sql
```

```
"SELECT `categories`.* FROM `categories` INNER JOIN `categories_profiles` ON `categories_profiles`.`category_id` = `categories`.`id` INNER JOIN `mobility_string_translations` `Category_slug_pt_br_string_translations` ON `Category_slug_pt_br_string_translations`.`key` = 'slug' AND `Category_slug_pt_br_string_translations`.`locale` = 'pt-BR' AND `Category_slug_pt_br_string_translations`.`translatable_type` = 'Category' AND `Category_slug_pt_br_string_translations`.`translatable_id` = `categories`.`id` WHERE `Category_slug_pt_br_string_translations`.`value` = 'the-title'"
```

But when I scope the `slug` into the hash, it does not works:

```
Category.i18n.joins(:category_profiles).where(categories: { slug: "the-title" }).to_sql
```

```
"SELECT `categories`.* FROM `categories` INNER JOIN `categories_profiles` ON `categories_profiles`.`category_id` = `categories`.`id` WHERE `categories`.`slug` = 'the-title'"
```

## Possible Fix

Maybe always trigger joins when `i18n` is used can solve the problem, mas if people to use it as `default_scope` will degrade performance.

I tried to use `join_mobility_model_translations` and `join_translations` but this methods are missing.

**Comment by shioyama** [comment]
This is not a bug, maybe a feature request, but realistically too complex to implement.

You should be able to do it though using `merge`, like this:

```ruby
Category.i18n.joins(:category_profiles).merge(CategoryProfile.where(slug: 'the-title'))
```

I'm assuming that you have a model `CategoryProfile` ? In any case `merge` should do it for you if you structure the query right.

**Comment by wbotelhos** [comment]
Hi @shioyama , thanks for your answer.

I did the following:

```
CategoryProfile
  .joins(:category)
  .merge(Category.i18n.where(slug: 'the-title'))
  .exists?(categories_profiles: { profile_id: id })
```

But was need to put the module on `CategoryProfile`:

```
NoMethodError: undefined method `mobility_modules' for #<ActiveRecord::Relation [#<CategoryProfile id: nil>]>
```

```
class CategoryProfile < ActiveRecord::Base
  extend Mobility
end
```

But just extend the module still does not works, was need set the callback:

```
translates :category_id, type: :string
```

But unfortunately this modules does not need translate, so I cannot do that.
Maybe activate the module with no need to set `translates` is a good way to solve it.

Thank you for your time.

**Comment by shioyama** [comment]
Sorry, this is really beyond the scope of an issue. You cannot use depth-2 hashes in your queries (so `exists?` will not work as you have it), you need to use `merge` instead, with `i18n` scope. If you are having problems doing that please post to stackoverflow.

**Comment by wbotelhos** [comment]
Sorry if it is not an issues:

```
NoMethodError: undefined method `mobility_modules' for #<ActiveRecord::Relation [#<CategoryProfile id: nil>]>
```

I'm using `merge` already as you said:

```
CategoryProfile
  .joins(:category)
  .merge(Category.i18n.where(slug: 'the-title'))
  .exists?(categories_profiles: { profile_id: id })
```

But i cannot translate the middle table just to work the join. I could do that on your other gem `globalize`.

Sorry about this issue.

**Comment by chamnap** [comment]
I'm having the same issue as @wbotelhos. It's not about `#exists?`.

```ruby
Achievement.joins(:activity).merge(Activity.18n.by_title(:asc))
```
This one works.

```
Achievement.joins(:activity).merge(Activity.i18n.by_title(:asc)).where(state: :pending)
> NoMethodError: undefined method `mobility_modules' for #<Achievement::ActiveRecord_Relation:0x00007fe979a71358>
```

```ruby
Achievement.where(state: :pending).joins(:activity).merge(Activity.i18n.by_title(:asc))
```
This one works.

Anything after merged with `i18n` scope doesn't work.

**Comment by shioyama** [comment]
I'm repeating, but **this is not an "issue"**. Querying through associations in a hash to `where` was never supported. I suggested using `merge` as an idea but I never even tried it, so it's not surprising that it might not actually work.

Questions on usage should go to stackoverflow, etc. not issues.

**Comment by shioyama** [comment]
I locked this but re-reading it, I'm unlocking it because the exception about `mobility_modules` might actually be a bug where the Mobility scope is applied to a non-Mobility model.

But I can't actually reproduce the exception myself.

Can someone provide:
- Mobility version (preferably use the latest 0.8.2)
- Code for models
- Code which produces the `mobility_modules` exception.

You should not be using nested hashes to `where` or other query methods on translated attributes, because this will not work (not supported by Mobility). But if you get this exception with normal queries, then this is a bug which can be fixed I think with a simple `respond_to?` in the Query plugin code.

**Comment by shioyama** [comment]
Or a failing spec would be even better :smile:

**Comment by shioyama** [comment]
@chamnap Yes but I cannot reproduce that error. If someone can give me an example with everything necessary to reproduce it, then I can fix it.

**Comment by shioyama** [comment]
Ok, I can reproduce it now using a nested hash:

If `Post` is an untranslated model, and `Comment` is a translated model with a translated attribute `content`, then this will blow up with the exception about `mobility_modules`:

```ruby
Post.joins(:comments).merge(Comment.i18n.where(content: "foo")).where(comments: { post_id: 5 })
```

This is a bug; this should work. It's easy to fix, I'll add a quick spec and release a new minor version this week.

**Comment by shioyama** [comment]
Fix released in 0.8.3.
